src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,357 @@
+                }
+            }
+
+            // Parse first 3 characters after a space as status code
+            // TODO: Parsing errors are treated as fatal. Find right behaviour
+            int statusCodeLength = 3;
+            int codeIndex = responseHeader.IndexOf(' ') + 1;
+            string strStatusCode = responseHeader.Substring(codeIndex, statusCodeLength);
+            response.StatusCode = (HttpStatusCode)(int.Parse(strStatusCode));
+
+            // The remaining string after the space is the reason phrase
+            response.ReasonPhrase = responseHeader.Substring(codeIndex + statusCodeLength + 1);
+
+            return true;
+        }
+
+        private static int CurlSocketCallback(
+            IntPtr handle,
+            curl_socket_t socketFd,
+            int socketAction,
+            IntPtr context,
+            IntPtr sockPtr)
+        {
+            int retVal = CURLMcode.CURLM_OK;
+
+            if (CurlPoll.CURL_POLL_REMOVE == socketAction)
+            {
+                return retVal;
+            }
+
+            bool isAdd = false;
+            if (IntPtr.Zero == sockPtr)
+            {
+                int result = Interop.libcurl.curl_easy_getinfo(handle, CURLINFO.CURLINFO_PRIVATE, out sockPtr);
+                if (result != CURLcode.CURLE_OK)
+                {
+                    throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result, false));
+                }
+                isAdd = true;
+            }
+
+            RequestCompletionSource state;
+            try
+            {
+                GCHandle stateHandle = GCHandle.FromIntPtr(sockPtr);
+                state = (RequestCompletionSource) stateHandle.Target;
+            }
+            catch
+            {
+                // Sometimes callbacks are invoked even after handle has been cleaned up
+                return retVal;
+            }
+
+            if (isAdd)
+            {
+                lock (state.SessionHandle)
+                {
+                    int result = Interop.libcurl.curl_multi_assign(state.SessionHandle, socketFd, sockPtr);
+                    if (result != CURLMcode.CURLM_OK)
+                    {
+                        throw new HttpRequestException(SR.net_http_client_execution_error,
+                            GetCurlException(result, true));
+                    }
+                }
+            }
+            else
+            {
+                int eventMask = CurlSelect.CURL_CSELECT_IN | CurlSelect.CURL_CSELECT_OUT;
+                CheckForCompletedTransfers(state.SessionHandle, socketFd, eventMask, sockPtr);
+            }
+
+            return retVal;
+        }
+
+        private static int CurlMultiTimerCallback(
+            IntPtr handle,
+            long timeoutInMilliseconds,
+            IntPtr context)
+        {
+            Task.Delay((int)timeoutInMilliseconds).ContinueWithStandard(context, (t, c) => CurlTimerElapsed((IntPtr)c));
```

Comment:
I was curious about this, so I looked at the libcurl docs, which includes in its discussion about the timeout argument:
```
A timeout_ms value of -1 means you should delete your timer.
```
so this doesn't appear to be doing the right thing...?